library/alloc/src/vec/mod.rs

```diff
@@ -1514,15 +1514,102 @@ impl<T, A: Allocator> Vec<T, A> {
+             * when we find it. */
+            write: usize,
+
+            /* The Vec that would need correction if `same_bucket` panicked */
+            vec: &'a mut Vec<T, A>,
+        }
+
+        impl<'a, T, A: core::alloc::Allocator> Drop for FillGapOnDrop<'a, T, A> {
+            fn drop(&mut self) {
+                /* This code gets executed either at the end of `dedup_by` or
+                 * when `same_bucket` panics */
+
+                /* SAFETY (if finishing successfully): self.read == len, so
+                 * no data is copied and length is set correctly */
+
+                /* SAFETY (if panicing): invariant guarantees that `read - write`
+                 * and `len - read` never overflow and that the copy is always
+                 * in-bounds. */
+                unsafe {
+                    let ptr = self.vec.as_mut_ptr();
+                    let len = self.vec.len();
+
+                    /* How many items were left when `same_bucket` paniced.
+                     * Basically vec[read..].len() */
+                    let items_left = len - self.read;
+
+                    /* Pointer to first item in vec[write..write+items_left] slice */
+                    let dropped_ptr = ptr.add(self.write);
+                    /* Pointer to first item in vec[read..] slice */
+                    let valid_ptr = ptr.add(self.read);
+
+                    /* Copy `vec[read..]` to `vec[write..write+items_left]`.
+                     * The slices can overlap, so `copy_nonoverlapping` cannot be used */
+                    ptr::copy(valid_ptr, dropped_ptr, items_left);
+
+                    /* How many items have been already dropped
+                     * Basically vec[read..write].len() */
+                    let dropped = self.read - self.write;
+
+                    self.vec.set_len(len - dropped);
+                }
+            }
+        }
+
+        let mut gap = FillGapOnDrop { read: 1, write: 1, vec: self };
+
+        let ptr = gap.vec.as_mut_ptr();
+
+        /* Drop items while going through Vec, it should be more efficient than
+         * doing slice partition_dedup + truncate */
+
+        /* SAFETY: Because of the invariant, read_ptr, prev_ptr and write_ptr
+         * are always in-bounds and read_ptr never aliases prev_ptr */
+        unsafe {
+            while gap.read < len {
+                let read_ptr = ptr.add(gap.read);
+                let prev_ptr = ptr.add(gap.write.wrapping_sub(1));
+
+                if same_bucket(&mut *read_ptr, &mut *prev_ptr) {
+                    /* We have found duplicate, drop it in-place */
+                    ptr::drop_in_place(read_ptr);
+                } else {
+                    let write_ptr = ptr.add(gap.write);
+
+                    /* Looks like doing just `copy` can be faster than
+                     * conditional `copy_nonoverlapping` */
+                    ptr::copy(read_ptr, write_ptr, 1);
```

Comment:
Will `read_ptr` ever equal `write_ptr` here? If not (which I believe holds), I don't see why a _conditional_ copy_nonoverlapping would be necessary.

Reply:
If no duplicate has been found yet, then `read_ptr == write_ptr`, that is the reason why either copy or conditional copy_nonoverlapping is needed. I have observed on my 2 laptops that the first option has overall better and more consistent performace

Reply:
> If no duplicate has been found yet, then read_ptr == write_ptr

> If no duplicate has been found yet, then read_ptr == write_ptr,

This doesn't sound right to me. There is an invariant above specified above that `gap.write-1 < gap.read` and `write_ptr` is always computed from `gap.write-1`, so it follows then that `write_ptr < read_ptr`, no?

Reply:
Ah, my bad! I confused it with the `prev_ptr`!